src/persist-types/src/arrow.rs

```diff
@@ -97,8 +98,7 @@ fn into_proto_with_type(data: &ArrayData, expected_type: Option<&DataType>) -> P
                 data.data_type(),
                 "actual type should match expected type"
             );
-            // TODO(bkirwi): return None here in a future release, assuming the types match
-            Some(expected.into_proto())
+            None
```

Comment:
Nice! The comment on this method seems a bit of out date though:

> We omit the data type _if_ if matches the expected type, to save space.

As-is we always omit the data type if `expected` is provided, it feels like checking against `data.data_type` would be nice, but I guess that is somewhat expensive? I'll leave it up to you!

Reply:
> but I guess that is somewhat expensive?

Specifically the trouble is that we end up checking equality N times for types at depth N, so it's easy to construct examples with quadratic performance. You could argue that deeply nested types are rare... but anyways it felt a bit silly to re-introduce quadratic overheads in an optimization that was specifically about removing quadratic overheads. :/